File: engine/src/main/java/de/hanno/hpengine/util/gui/container/ReloadableScrollPane.java
package de.hanno.hpengine.util.gui.container;

import javax.swing.*;
import java.awt.*;

public class ReloadableScrollPane extends JScrollPane {

    private final Component child;

    public <TYPE extends Component> ReloadableScrollPane(TYPE component) {
        super(component);
        if(component == null) {
            throw new IllegalArgumentException("Don't pass null child");
        }
        this.child = component;
    }

    public void reload() {
        setViewportView(child);
        child.revalidate();
        child.repaint();
        this.getRootPane().invalidate();
        this.getRootPane().revalidate();
        this.getRootPane().repaint();
    }
}

File: engine/src/main/java/de/hanno/hpengine/util/gui/container/ReloadableScrollPane.kt
package de.hanno.hpengine.util.gui.container

import java.awt.Component
import javax.swing.JScrollPane
import javax.swing.ScrollPaneConstants

class ReloadableScrollPane(val child: Component, fixedWidth: Int? = null) : JScrollPane(child) {
    init {
        fixedWidth?.let { preferredSize.width = it }
        setHorizontalScrollBarPolicy(ScrollPaneConstants.HORIZONTAL_SCROLLBAR_NEVER)
        setVerticalScrollBarPolicy(ScrollPaneConstants.VERTICAL_SCROLLBAR_AS_NEEDED)
    }

    fun reload() {
        setViewportView(child)
        child.revalidate()
        child.repaint()

        rootPane.invalidate()
        rootPane.revalidate()
        rootPane.repaint()
    }
}